### ISSUE TITLE
Fail to terminate pvc and pod using ceph block storage

### ISSUE THREAD

**Issue by aisensiy** [rook/rook#2261]
I searched "terminating" "dead lock" but no specific solution fit for me. I don't have much experience working with ceph. I tried to access kubelet logs, osd logs, mon logs, mgr logs and tried to restart mon pod, mrg pod and kubelet service...but all of these did not helpful for this issue.

**Is this a bug report or feature request?**
* Bug Report

**Deviation from expected behavior:**

Delete pod and then the pod get stuck in `Terminating...` status and so is the pvc.

**Expected behavior:**

Delete successfully.

**How to reproduce it (minimal and precise):**

Create a pvc like this:

```
apiVersion: v1
kind: PersistentVolumeClaim
metadata:
  annotations:
    volume.beta.kubernetes.io/storage-provisioner: ceph.rook.io/block
  finalizers:
  - kubernetes.io/pvc-protection
  labels:
    app: wordpress
  name: wp-pv-claim
  namespace: default
  resourceVersion: "900085"
  selfLink: /api/v1/namespaces/default/persistentvolumeclaims/wp-pv-claim
  uid: 2de72ebc-de71-11e8-93bd-02e8e9d8fba8
spec:
  accessModes:
  - ReadWriteOnce
  - ReadOnlyMany
  resources:
    requests:
      storage: 20Gi
  storageClassName: rook-ceph-block
  volumeName: pvc-2de72ebc-de71-11e8-93bd-02e8e9d8fba8
status:
  accessModes:
  - ReadWriteOnce
  - ReadOnlyMany
  capacity:
    storage: 20Gi
  phase: Bound
```

Then create a pod to use this pvc:

```
apiVersion: v1
kind: Pod
metadata:
  name: busybox2
spec:
  volumes:
  - name: workdir
    persistentVolumeClaim:
      claimName: wp-pv-claim
      readOnly: true
  containers:
  - name: busybox-test
    image: busybox
    resources:
      limits:
        cpu: 100m
        memory: 100Mi
    command: ["/bin/sh"]
    args: ["-c", "sleep 3600"]
    volumeMounts:
    - name: workdir
      mountPath: /var/data
      readOnly: true
```

Then the server has something wrong and I restart the server. Then all the `rook-ceph` and `rook-ceph-system` pods works well but when I `kubectl delete pod busybox2` it suck into `Terminating` status...

**Environment**:
* OS (e.g. from /etc/os-release):

```
NAME="Ubuntu"
VERSION="16.04.4 LTS (Xenial Xerus)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 16.04.4 LTS"
VERSION_ID="16.04"
HOME_URL="http://www.ubuntu.com/"
SUPPORT_URL="http://help.ubuntu.com/"
BUG_REPORT_URL="http://bugs.launchpad.net/ubuntu/"
VERSION_CODENAME=xenial
UBUNTU_CODENAME=xenial
```

* Kernel (e.g. `uname -a`):

`Linux xxxx.compute.internal 4.4.0-138-generic #164-Ubuntu SMP Tue Oct 2 17:16:02 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux`

* Cloud provider or hardware configuration:

aws ec2 mediam

* Rook version (use `rook version` inside of a Rook Pod):

```
rook: v0.8.0-263.g9ba4fcc
```

* Kubernetes version (use `kubectl version`):

```
Client Version: version.Info{Major:"1", Minor:"10", GitVersion:"v1.10.2", GitCommit:"81753b10df112992bf51bbc2c2f85208aad78335", GitTreeState:"clean", BuildDate:"2018-05-12T04:12:12Z", GoVersion:"go1.9.6", Compiler:"gc", Platform:"darwin/amd64"}
Server Version: version.Info{Major:"1", Minor:"11", GitVersion:"v1.11.3", GitCommit:"a4529464e4629c21224b3d52edfe0ea91b072862", GitTreeState:"clean", BuildDate:"2018-09-09T17:53:03Z", GoVersion:"go1.10.3", Compiler:"gc", Platform:"linux/amd64"}
```

* Kubernetes cluster type (e.g. Tectonic, GKE, OpenShift):

```
Bare Metal, created by kubeadm
```

* Storage backend status (e.g. for Ceph use `ceph health` in the [Rook Ceph toolbox](https://rook.io/docs/Rook/master/toolbox.html)):

```
  cluster:
    id:     60a2875f-d962-44cd-a22d-7251b982895c
    health: HEALTH_OK
 
  services:
    mon: 3 daemons, quorum b,a,c
    mgr: a(active)
    mds: openbayes-data-cache-1/1/1 up  {0=openbayes-data-cache-b=up:active}, 1 up:standby-replay
    osd: 3 osds: 3 up, 3 in
 
  data:
    pools:   5 pools, 500 pgs
    objects: 81 objects, 132 MB
    usage:   19180 MB used, 276 GB / 294 GB avail
    pgs:     500 active+clean
 
  io:
    client:   852 B/s rd, 1 op/s rd, 0 op/s wr
```

**Comment by hydratim** [comment]
I have the same issue:
K8s:
 ```
Client Version: version.Info{Major:"1", Minor:"13", GitVersion:"v1.13.0", GitCommit:"ddf47ac13c1a9483ea035a79cd7c10005ff21a6d", GitTreeState:"clean", BuildDate:"2018-12-03T21:04:45Z", GoVersion:"go1.11.2", Compiler:"gc", Platform:"linux/amd64"}
Server Version: version.Info{Major:"1", Minor:"11", GitVersion:"v1.11.5", GitCommit:"753b2dbc622f5cc417845f0ff8a77f539a4213ea", GitTreeState:"clean", BuildDate:"2018-11-26T14:31:35Z", GoVersion:"go1.10.3", Compiler:"gc", Platform:"linux/amd64"}
```
Uname -a:
```
Linux uk-lon-49 4.15.0-43-generic #46~16.04.1-Ubuntu SMP Fri Dec 7 13:31:08 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux
```
rook version:
```
rook: v0.9.0
```
ceph status:
```
  cluster:
    id:     a1a8eef4-0f3f-4d9c-9acd-22ac6bd219b3
    health: HEALTH_ERR
            2 modules have failed
            Degraded data redundancy: 1238/1857 objects degraded (66.667%), 99 pgs degraded, 100 pgs undersized
            mons a,b,c are low on available space
 
  services:
    mon: 3 daemons, quorum a,b,c
    mgr: a(active)
    osd: 1 osds: 1 up, 1 in
 
  data:                                                                                                                                                                                                                                                                        
    pools:   1 pools, 100 pgs                                                                                                                                                                                                                                                  
    objects: 619  objects, 2.3 GiB                                                                                                                                                                                                                                             
    usage:   14 GiB used, 5.5 GiB / 20 GiB avail                                                                                                                                                                                                                               
    pgs:     1238/1857 objects degraded (66.667%)                                                                                                                                                                                                                              
             99 active+undersized+degraded                                                                                                                                                                                                                                     
             1  active+undersized                                                                                                                                                                                                                                              
```                                                                                                                                                                                                                                                                          
ceph osd status:
```                                                                                                                                                                                                                                          
+----+----------------------------------+-------+-------+--------+---------+--------+---------+-----------+                                                                                                                                                                    
| id |               host               |  used | avail | wr ops | wr data | rd ops | rd data |   state   |                                                                                                                                                                    
+----+----------------------------------+-------+-------+--------+---------+--------+---------+-----------+                                                                                                                                                                    
| 0  | rook-ceph-osd-0-7bf499d4d9-dxrhz | 14.0G | 5657M |    0   |     0   |    0   |     0   | exists,up |                                                                                                                                                                    
+----+----------------------------------+-------+-------+--------+---------+--------+---------+-----------+                                                                                                                                                                    
```
ceph df:
```                                                                                                                                                                                                           
GLOBAL:                                                                                                                                                                                                                                                                        
    SIZE       AVAIL       RAW USED     %RAW USED                                                                                                                                                                                                                              
    20 GiB     5.5 GiB       14 GiB         71.75                                                                                                                                                                                                                              
POOLS:                                                                                                                                                                                                                                                                         
    NAME                            ID     USED        %USED     MAX AVAIL     OBJECTS                                                                                                                                                                                         
    docker-registry-replicapool     14     2.3 GiB     33.16       1.5 GiB         619    
```

**Comment by hydratim** [comment]
When I re-insert the sc I get:
```
flexdriver: Unmount volume at mount dir /var/lib/kubelet/pods/a1931d2b-15bf-11e9-b425-005056a52db7/volumes/ceph.rook.io~rook-ceph-system/pvc-a192d41c-15bf-11e9-b425-005056a52db7 failed: failed to get persistent volume pvc-a192d41c-15bf-11e9-b425-005056a52db7: persistentvolumes "pvc-a192d41c-15bf-11e9-b425-005056a52db7" not found
```

**Comment by hydratim** [comment]
Okay so from my tests this seems to be triggered by any helm chart (or other automated install platform) that contains a cephpool and/or a StorageController.

This is likely because of a lacking pre-condition in the rook/ceph drivers (for Storage Controller, Pool, etc) that should block the finalizer untill any pvc's or other "child" objects have been removed/cleaned up. 

If K8s first removes the SC and/or Pool and then tries to remove the Pods and/or PVCs - The kubelet will  then fail to umount the RBD which results in the Pod failing to terminate pending the umount finalizer (that will never return 0).

**Comment by galexrt** [comment]
@aisensiy @hydratim What `replicated.size` are you using in your CephBlockPool?

**Comment by aisensiy** [comment]
Mine is `2`

**Comment by hydratim** [comment]
IIRC I was using the defaults from the helm charts, however I don't recall precisely.

**Comment by namku** [comment]
I have the same issue when I shutdown one node de pod stuck in terminating state..

**Comment by namku** [comment]
Is it solved with rock v1.1 beta?
